File: java-learning-path/03-large-projects/library-management/src/main/java/com/example/javaguide/library/service/LibraryService.java
package com.example.javaguide.library.service;

import com.example.javaguide.library.model.*;
import com.example.javaguide.library.repository.BookRepository;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 * Service layer for the Library Management System.
 * Orchestrates operations between repositories and domain services,
 * managing books, members, and the borrowing/returning lifecycle.
 */
public class LibraryService {
    private static LibraryService instance;
    private final BookRepository bookRepository;
    private final LoanService loanService;
    private final MemberService memberService;
    
    private LibraryService() {
        this.bookRepository = new BookRepository("data/books.json");
        this.loanService = new LoanService();
        this.memberService = new MemberService();
    }
    
    /**
     * Pulls sequential contexts.
     */
    public static synchronized LibraryService getInstance() {
        if (instance == null) {
            instance = new LibraryService();
        }
        return instance;
    }
    
    // Book operations
    public void addBook(Book book) {
        bookRepository.save(book);
    }
    
    public Optional<Book> findBook(String isbn) {
        return bookRepository.findByIsbn(isbn);
    }
    
    public List<Book> searchBooksByTitle(String title) {
        return bookRepository.findByTitle(title);
    }
    
    public List<Book> searchBooksByAuthor(String author) {
        return bookRepository.findByAuthor(author);
    }
    
    public List<Book> getAllBooks() {
        return bookRepository.findAll();
    }
    
    // Member operations
    public void registerMember(Member member) {
        memberService.addMember(member);
    }
    
    public Optional<Member> findMember(String memberId) {
        return memberService.findById(memberId);
    }
    
    // Loan operations
    /**
     * Processes a book borrowing request.
     */
    public String borrowBook(String memberId, String isbn) {
        Member member = memberService.findById(memberId)
                .orElseThrow(() -> new IllegalArgumentException("Member not found"));
        
        Book book = bookRepository.findByIsbn(isbn)
                .orElseThrow(() -> new IllegalArgumentException("Book not found"));
        
        if (!book.isAvailable()) {
            throw new IllegalStateException("Book not available");
        }
        
        if (!member.canBorrowMore()) {
            throw new IllegalStateException("Member has reached borrowing limit");
        }
        
        String loanId = UUID.randomUUID().toString();
        Loan loan = new Loan(loanId, memberId, isbn);
        
        book.setStatus(Book.BookStatus.BORROWED);
        bookRepository.save(book);
        
        member.borrowBook(isbn);
        memberService.updateMember(member);
        
        loanService.addLoan(loan);
        
        return loanId;
    }
    
    /**
     * Processes a book returning request.
     */
    public void returnBook(String loanId) {
        Loan loan = loanService.findById(loanId)
                .orElseThrow(() -> new IllegalArgumentException("Loan not found"));
        
        Book book = bookRepository.findByIsbn(loan.getIsbn())
                .orElseThrow(() -> new IllegalArgumentException("Book not found"));
        
        Member member = memberService.findById(loan.getMemberId())
                .orElseThrow(() -> new IllegalArgumentException("Member not found"));
        
        loan.returnBook();
        loanService.updateLoan(loan);
        
        book.setStatus(Book.BookStatus.AVAILABLE);
        bookRepository.save(book);
        
        member.returnBook(loan.getIsbn());
        memberService.updateMember(member);
    }
}

// Supporting classes
class LoanService {
    private final java.util.Map<String, Loan> loans = new java.util.HashMap<>();
    
    public void addLoan(Loan loan) {
        loans.put(loan.getLoanId(), loan);
    }
    
    public Optional<Loan> findById(String loanId) {
        return Optional.ofNullable(loans.get(loanId));
    }
    
    public void updateLoan(Loan loan) {
        loans.put(loan.getLoanId(), loan);
    }
}

class MemberService {
    private final java.util.Map<String, Member> members = new java.util.HashMap<>();
    
    public void addMember(Member member) {
        members.put(member.getMemberId(), member);
    }
    
    public Optional<Member> findById(String memberId) {
        return Optional.ofNullable(members.get(memberId));
    }
    
    public void updateMember(Member member) {
        members.put(member.getMemberId(), member);
    }
}

File: python-learning-path/03-large-projects/library-management/services/library_service.py
"""
Library Service Module

Provides the main LibraryService class that orchestrates all library
operations including book management, member registration, and loan handling.
Uses the Singleton pattern to ensure a single instance across the application.
"""
from typing import Optional
import uuid
from ..models.book import Book, BookStatus
from ..models.member import Member, MembershipType
from ..models.loan import Loan, LoanStatus
from ..repository.book_repository import BookRepository


class LibraryService:
    """
    Main service class for library operations.

    Implements the Singleton pattern to ensure only one instance
    manages all library operations throughout the application.
    """
    _instance = None

    def __new__(cls):
        """Singleton implementation - ensures single instance."""
        if cls._instance is None:
            cls._instance = super(LibraryService, cls).__new__(cls)
            cls._instance._init_service()
        return cls._instance

    def _init_service(self):
        """Initialize internal state (called once on first instantiation)."""
        self.book_repo = BookRepository("data/books.json")
        self.members: dict[str, Member] = {}
        self.loans: dict[str, Loan] = {}

    def add_book(self, book: Book):
        """Add a new book to the library collection."""
        self.book_repo.save(book)

    def register_member(self, member: Member):
        """Register a new member in the library system."""
        self.members[member.member_id] = member

    def borrow_book(self, member_id: str, isbn: str) -> str:
        """
        Process a book borrowing transaction.

        Returns the loan_id. Raises ValueError if validation fails.
        """
        member = self.members.get(member_id)
        if not member: raise ValueError("Member not found")

        book = self.book_repo.find_by_isbn(isbn)
        if not book: raise ValueError("Book not found")

        if book.status != BookStatus.AVAILABLE:
            raise ValueError("Book is not available")

        if not member.can_borrow_more():
            raise ValueError("Member reached borrowing limit")

        # Create Loan record with unique 8-char ID
        loan_id = str(uuid.uuid4())[:8]
        loan = Loan(loan_id, member_id, isbn)
        self.loans[loan_id] = loan

        # Update book status and member's borrowed list
        book.status = BookStatus.BORROWED
        self.book_repo.save(book)
        member.borrow_book(isbn)

        return loan_id

    def return_book(self, loan_id: str):
        """Process a book return. Raises ValueError if loan not found."""
        loan = self.loans.get(loan_id)
        if not loan: raise ValueError("Loan not found")

        book = self.book_repo.find_by_isbn(loan.isbn)
        member = self.members.get(loan.member_id)

        # Update all statuses
        loan.status = LoanStatus.RETURNED
        book.status = BookStatus.AVAILABLE
        self.book_repo.save(book)
        member.return_book(loan.isbn)
